src/Sylius/Bundle/ApiBundle/Validator/Constraints/EligibleProductVariantToCartValidator.php

```diff
@@ -107,4 +122,23 @@ public function validate($value, Constraint $constraint)
             );
         }
     }
+
+    private function assignProductVariantAndProductVariantCode($value): void
```

Comment:
I wonder if it would be better to add the second validator for `ChangeItemQuantityInCart` command 🤔

Reply:
I thought about it, but IMO both validators are responsible for validation adding items with quantity. Is easier remember about one validator if would be something to fix. Maybe I should to add a trait or not final class with related common part.

Reply:
On the other hand, if one would like to customize behavior of adding to cart validator, but not the change items quantity, they would have a hard time. 
Why not inverse the logic? Right now I see ~10 paths going throughout this class. It's way too much for regular service. On the other hand, there are only three types of validation errors. If you would extract common checking logic to some services like:
ProductVariantPurchaseAvailibilityChecker - to check if stock is available
ProductVariantEnebilityChecker - to check if variant exist/is turned on
ProductEnebilityChecker - to check if product exist/is turned on

The first two could have even the same interface, but that would mean that we are diffusing the meaning of each service - so I would rather even consider packing them together and leaving only two types of communication from here. Or just stick to the interface for each of the above-mentioned checkers.

In addition, it will be waaaay easier to test it with spec. Another side effect? Adding new checking logic will be piece of cake